usr/share/rear/prep/GNU/Linux/210_include_dhclient.sh

```diff
@@ -4,114 +4,123 @@
 # copy the binaries and config files that we require to use dhclient/dhcpcd
 # on the rescue image (IPv4/IPv6)
 
-define_dhclients_variable()
-{   local x
-    dhclients=()
-    for x in "${DHCLIENT_BIN##*/}" \
-        "${DHCLIENT6_BIN##*/}" \
-        dhcpcd dhclient \
-        dhcp6c dhclient6 ;
-    do
-        [ "x$x" == x ] && continue
-        for d in ${dhclients[@]} ; do
-            [ "x$d" = "x$x" ] && continue 2
-        done
-        dhclients=(${dhclients[@]} "$x")
-    done
-    dhclients=${dhclients[@]}
-}
+# Respect an explicit rejection
+if is_false "$USE_DHCLIENT"; then
+    Log "Disabling DHCP client on the rescue system (USE_DHCLIENT='$USE_DHCLIENT')"
+    return 0
+fi
 
-dhcp_interfaces_active() {
-    local my_dhclients
-    my_dhclients=${dhclients}
-    my_dhclients=${dhclients// /|}
-    ps -e | grep -qEs "[ /]($my_dhclients)"
+# Array of known DHCP clients, which must also be known to skel/default/etc/scripts/system-setup.d/58-start-dhclient.sh
+local dhcp_clients=(dhcpcd dhclient dhcp6c dhclient6)
+
+# DHCLIENT_BIN, if set, specifies the DHCP client binary to use on the rescue system.
+# Must be one of the binaries listed in $dhcp_clients. Default (empty): try to auto-detect.
+# FIXME: DHCLIENT_BIN could be documented in default.conf as needed.
+if [[ -n "$DHCLIENT_BIN" ]]; then
+    has_binary "$DHCLIENT_BIN" || Error "DHCLIENT_BIN='$DHCLIENT_BIN' but such a binary could not be found"
+    [[ "${dhcp_clients[*]}" == *" $DHCLIENT_BIN "* ]] || Error "DHCLIENT_BIN='$DHCLIENT_BIN' is not among known DHCP clients (${dhcp_clients[*]})"
+fi
+
+
+function dhcp_client_is_active() {
+    # Strategy 1: Check if any DHCP client process is running (does not need to be the one specified by DHCLIENT_BIN)
+    local dhcp_clients_regexp="${dhcp_clients[*]}"
+    dhcp_clients_regexp=${dhcp_clients_regexp// /|}
+    ps -e | grep -qEs "[ /]($dhcp_clients_regexp)"
     if [ $? -eq 0 ]; then
-        # if we find a dhcp client being active we automatically
-        # include DHCP CLIENT support in Relax-and-Recover
-        Log "Running DHCP client found, enabling USE_DHCLIENT"
-        USE_DHCLIENT=y
+        Log "Detected an active DHCP client process"
+        return 0
     fi
-}
 
-define_dhclient_bins()
-{   # purpuse is to define which binaries are being used on this system
-    # other dhcp clients should be hard-coded in the /etc/rear/[site|local].conf file
-    case ${1##*/} in
-        dhcpcd) DHCLIENT_BIN=dhcpcd ;;
-        dhcp6c) DHCLIENT6_BIN=dhcp6c ;;
-        dhclient) DHCLIENT_BIN=dhclient ;;
-        dhclient6) DHCLIENT6_BIN=dhclient6 ;;
-    esac
+    # Strategy 2: Check if systemd-networkd is active and configured to use its built-in DHCP client
+    if type systemctl &>/dev/null && systemctl is-active --quiet systemd-networkd; then
```

Comment:
Personally I would prefer to see error messages in the log via
```
if type systemctl 1>/dev/null && systemctl ...
```
which makes it much easier for us when a user reports
an issue related to DHCP setup in his recovery system
and provides his "rear -D mkrescue/mkbackup" log
when we therein could see how things fail e.g. like
```
+ type systemctl
-bash: type: systemctl: not found
```
Perhaps that is OK because he has an old system
without systemd (e.g. SLES11)
perhaps that is not OK because he should have systemd
and then we could see that his systemd setup looks broken.

Reply:
OK, I've changed the line to `has_binary` though it's not my preference. These "not found" messages during normal operations make ReaR look broken in the log when it's not. They also don't provide information about their actual point of invocation so if you have many of these checks it's hard to figure out what was intended and what might be an error.

Reply:
Hm, after looking at the log I've actually decided not to change this. It just looks too broken and if there's no `systemctl` we can be sure that there is no working `systemd` installation at all.

Feel free to change afterwards if you like.

Reply:
Perfectly fine with me.
It is your code so you make the final decisions
(except there would be real errors in the code)
but this here is only a matter of personal preferences.
And I fully agree that there is a dilemman with such messages in the log.
On the one habd too many false positives
on the other hand to little info.
The only "solution" as far as I can imagine would be
even more elaborated code with lots of explicit error checking code
but that additional error checking code can make the actual code
almost invisible within all those error checking code around it.